sdk/src/main/scala/com/horizen/chain/ActiveChain.scala

```diff
@@ -31,88 +33,123 @@ final class ActiveChain private(private val sidechainCache: ElementsChain[Modifi
+  def mcHeadersHeightByMcHash(mainchainHeaderHash: MainchainHeaderHash): Option[Int] = mainchainHeadersCache.heightById(mainchainHeaderHash).map(_ + mainchainCreationBlockHeightDifference)
+
+  def mcRefDataHeightByMcHash(mainchainHeaderHash: MainchainHeaderHash): Option[Int] = mainchainReferenceDataCache.heightById(mainchainHeaderHash).map(_ + mainchainCreationBlockHeightDifference)
 
-  def mcIdByMcHeight(mcHeight: Int): Option[MainchainBlockReferenceId] = mainchainCache.idByHeight(mcHeight - mainchainCreationBlockHeightDifference)
+  def mcHashByMcHeight(mcHeight: Int): Option[MainchainHeaderHash] = mainchainHeadersCache.idByHeight(mcHeight - mainchainCreationBlockHeightDifference)
 
   // Active chain store
-  def heightOfMc: Int = mainchainCache.height + mainchainCreationBlockHeightDifference
+  def heightOfMcHeaders: Int = mainchainHeadersCache.height + mainchainCreationBlockHeightDifference
+
+  def heightOfMcReferencesData: Int = mainchainReferenceDataCache.height + mainchainCreationBlockHeightDifference
+
+  def mcHeaderMetadataByMcHash(mainchainHeaderHash: MainchainHeaderHash): Option[MainchainHeaderMetadata] = mainchainHeadersCache.dataById(mainchainHeaderHash)
 
-  def mcBlockReferenceDataByMcId(mainChainReferenceId: MainchainBlockReferenceId): Option[MainchainBlockReferenceData] = mainchainCache.dataById(mainChainReferenceId)
+  def mcReferenceDataMetadataByMcHash(mainchainHeaderHash: MainchainHeaderHash): Option[MainchainHeaderMetadata] = mainchainReferenceDataCache.dataById(mainchainHeaderHash)
 
   // Mixed data retrieval
-  def heightByMcId(mcId: MainchainBlockReferenceId): Option[Int] = {
-    mainchainCache.dataById(mcId).map(_.sidechainHeight)
+  def heightByMcHeader(mainchainHeaderHash: MainchainHeaderHash): Option[Int] = {
+    mainchainHeadersCache.dataById(mainchainHeaderHash).map(_.sidechainHeight)
   }
 
-  def idByMcId(mcId: MainchainBlockReferenceId): Option[ModifierId] = {
-    heightByMcId(mcId).flatMap(sidechainCache.idByHeight)
+  def idByMcHeader(mainchainHeaderHash: MainchainHeaderHash): Option[ModifierId] = {
+    heightByMcHeader(mainchainHeaderHash).flatMap(sidechainCache.idByHeight)
   }
 
-  def setBestBlock(newBestId: ModifierId, newBestData: SidechainBlockInfo, mainchainParentId: Option[MainchainBlockReferenceId]): Unit = {
+  def heightByMcReferenceData(mainchainReferenceDataHeaderHash: MainchainHeaderHash): Option[Int] = {
+    mainchainReferenceDataCache.dataById(mainchainReferenceDataHeaderHash).map(_.sidechainHeight)
+  }
+
+  def idByMcReferenceData(mainchainReferenceDataHeaderHash: MainchainHeaderHash): Option[ModifierId] = {
+    heightByMcReferenceData(mainchainReferenceDataHeaderHash).flatMap(sidechainCache.idByHeight)
+  }
+
+  def setBestBlock(newBestId: ModifierId, newBestData: SidechainBlockInfo, mainchainParentHashOpt: Option[MainchainHeaderHash]): Unit = {
```

Comment:
I didn't add it, because `mainchainParentHashOpt` is retrieved from newBest `SidechainBlock` first MainchainHeader parent hash. But `MainchainReferenceData` has no info about the parent. The only way to get it outside in History is from `ActiveChain` itself, so no sense. I just trust to the validation rules processed before.

Reply:
according to discussion, proper addition checks were added below.